File: tools/fuzz/seed_corpus.mjs
#!/usr/bin/env node
// Seeds tests/fuzz/corpus/<target>/ from repository fixtures so libFuzzer starts
// with a coverage map that reflects realistic SFC and template shapes. The
// seed file content is reproducible from the repo (the .vue and template
// fixtures are all in git), so the grown corpus does not need to be
// checked in — CI caches it instead.
//
// Targets currently seeded:
//   - sfc_parse: whole .vue files
//   - template_lexer: contents of <template>...</template> blocks
//   - js_ts_expression: script snippets and template interpolation expressions
//   - css_parse: contents of <style>...</style> blocks
//   - template_compile: contents of <template>...</template> blocks
//
// Usage: node tools/fuzz/seed_corpus.mjs

import { createHash } from "node:crypto";
import { globSync, mkdirSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const here = dirname(fileURLToPath(import.meta.url));
const repoRoot = resolve(here, "..", "..");
const corpusRoot = join(repoRoot, "tests", "fuzz", "corpus");

const VUE_GLOBS = [
  "tests/fixtures/**/*.vue",
  "tests/_fixtures/_projects/**/*.vue",
  "playground/src/**/*.vue",
  "playground/e2e/**/*.vue",
];

function findVueFiles() {
  return VUE_GLOBS.flatMap((pattern) =>
    globSync(pattern, { cwd: repoRoot }).map((relativePath) => join(repoRoot, relativePath)),
  );
}

function hash(content) {
  return createHash("sha1").update(content).digest("hex").slice(0, 16);
}

function resetCorpus(target) {
  const dir = join(corpusRoot, target);
  rmSync(dir, { force: true, recursive: true });
  mkdirSync(dir, { recursive: true });
  return dir;
}

function writeSeed(dir, content) {
  if (content.length === 0) return;
  writeFileSync(join(dir, hash(content)), content);
}

function extractTemplateBlock(source) {
  // The fuzz target consumes raw template source (the content between
  // <template> and </template>, with all attributes stripped). This pulls
  // the *first* template block from each SFC. Skipped if the file has no
  // template (e.g. script-only fixtures).
  const open = source.match(/<template\b[^>]*>/);
  if (!open) return null;
  const start = open.index + open[0].length;
  const closeIdx = source.indexOf("</template>", start);
  if (closeIdx === -1) return null;
  return source.slice(start, closeIdx);
}

function extractBlocks(source, tagName) {
  const blocks = [];
  const pattern = new RegExp(`<${tagName}\\b[^>]*>([\\s\\S]*?)</${tagName}>`, "gi");
  for (const match of source.matchAll(pattern)) {
    blocks.push(match[1]);
  }
  return blocks;
}

function extractInterpolations(template) {
  const expressions = [];
  for (const match of template.matchAll(/\{\{([\s\S]*?)\}\}/g)) {
    expressions.push(match[1].trim());
  }
  return expressions;
}

function main() {
  const sfcDir = resetCorpus("sfc_parse");
  const templateLexerDir = resetCorpus("template_lexer");
  const jsTsExpressionDir = resetCorpus("js_ts_expression");
  const cssParseDir = resetCorpus("css_parse");
  const templateDir = resetCorpus("template_compile");

  const files = findVueFiles();
  let sfcCount = 0;
  let templateCount = 0;
  let expressionCount = 0;
  let styleCount = 0;
  for (const path of files) {
    const content = readFileSync(path, "utf8");
    writeSeed(sfcDir, content);
    sfcCount += 1;

    const template = extractTemplateBlock(content);
    if (template != null) {
      writeSeed(templateLexerDir, template);
      writeSeed(templateDir, template);
      for (const expression of extractInterpolations(template)) {
        writeSeed(jsTsExpressionDir, expression);
        expressionCount += 1;
      }
      templateCount += 1;
    }

    for (const script of extractBlocks(content, "script")) {
      writeSeed(jsTsExpressionDir, script);
      expressionCount += 1;
    }

    for (const style of extractBlocks(content, "style")) {
      writeSeed(cssParseDir, style);
      styleCount += 1;
    }
  }

  process.stdout.write(
    `Seeded ${sfcCount} sfc_parse entries, ${templateCount} template entries, ${expressionCount} JS/TS expression entries, and ${styleCount} CSS entries from ${files.length} fixtures.\n`,
  );
}

main();

File: tools/commands/ci/fuzz/seed_corpus.rs
#!/usr/bin/env rust-script
//! ```cargo
//! [dependencies]
//! glob = "0.3"
//! regex = "1"
//! serde = { version = "1", features = ["derive"] }
//! serde_json = "1"
//! sha1 = "0.10"
//!
//! [package]
//! edition = "2024"
//! ```

#[path = "../../../rust/common.rs"]
mod common;

use regex::Regex;
use sha1::{Digest, Sha1};
use std::{
    fs,
    path::{Path, PathBuf},
    process::ExitCode,
};

const VUE_GLOBS: &[&str] = &[
    "tests/fixtures/**/*.vue",
    "tests/_fixtures/_projects/**/*.vue",
    "playground/src/**/*.vue",
    "playground/e2e/**/*.vue",
];

const FOLIO_GLOBS: &[&str] = &[
    "crates/vize_davinci/tests/fixtures/**/*.folio",
    "crates/vize_s2/tests/fixtures/**/*.folio",
];

fn main() -> ExitCode {
    common::main_result(run())
}

fn run() -> Result<(), String> {
    let root = common::repo_root()?;
    let corpus_root = root.join("tests/fuzz/corpus");

    let sfc_dir = reset_corpus(&corpus_root, "sfc_parse")?;
    let template_lexer_dir = reset_corpus(&corpus_root, "template_lexer")?;
    let expression_dir = reset_corpus(&corpus_root, "js_ts_expression")?;
    let css_dir = reset_corpus(&corpus_root, "css_parse")?;
    let template_dir = reset_corpus(&corpus_root, "template_compile")?;
    let s1_lowering_dir = reset_corpus(&corpus_root, "s1_lowering")?;
    let folio_dir = reset_corpus(&corpus_root, "folio_parse")?;

    let mut folio_count = 0usize;
    for file in glob_files(&root, FOLIO_GLOBS)? {
        write_seed(
            &folio_dir,
            &fs::read(&file).map_err(|error| format!("cannot read {}: {error}", file.display()))?,
        )?;
        folio_count += 1;
    }

    let vue_files = glob_files(&root, VUE_GLOBS)?;
    let script_block = Regex::new(r"(?is)<script\b[^>]*>(.*?)</script>").unwrap();
    let style_block = Regex::new(r"(?is)<style\b[^>]*>(.*?)</style>").unwrap();
    let interpolation = Regex::new(r"(?s)\{\{(.*?)\}\}").unwrap();

    let mut sfc_count = 0usize;
    let mut template_count = 0usize;
    let mut expression_count = 0usize;
    let mut style_count = 0usize;

    for file in &vue_files {
        let content = fs::read_to_string(file)
            .map_err(|error| format!("cannot read {}: {error}", file.display()))?;
        write_seed(&sfc_dir, content.as_bytes())?;
        sfc_count += 1;

        if let Some(template) = extract_template_block(&content) {
            write_seed(&template_lexer_dir, template.as_bytes())?;
            write_seed(&template_dir, template.as_bytes())?;
            write_seed(&s1_lowering_dir, template.as_bytes())?;
            for capture in interpolation.captures_iter(template) {
                write_seed(&expression_dir, capture[1].trim().as_bytes())?;
                expression_count += 1;
            }
            template_count += 1;
        }

        for capture in script_block.captures_iter(&content) {
            write_seed(&expression_dir, capture[1].as_bytes())?;
            expression_count += 1;
        }

        for capture in style_block.captures_iter(&content) {
            write_seed(&css_dir, capture[1].as_bytes())?;
            style_count += 1;
        }
    }

    println!(
        "Seeded {sfc_count} sfc_parse entries, {template_count} template entries (template_lexer/template_compile/s1_lowering), {expression_count} JS/TS expression entries, {style_count} CSS entries, and {folio_count} folio pages from {} fixtures.",
        vue_files.len() + folio_count
    );
    Ok(())
}

fn glob_files(root: &Path, patterns: &[&str]) -> Result<Vec<PathBuf>, String> {
    let mut files = Vec::new();
    for pattern in patterns {
        let absolute = root.join(pattern);
        let pattern = absolute
            .to_str()
            .ok_or_else(|| format!("non-utf8 glob pattern: {}", absolute.display()))?;
        for entry in glob::glob(pattern).map_err(|error| error.to_string())? {
            files.push(entry.map_err(|error| error.to_string())?);
        }
    }
    files.sort();
    Ok(files)
}

fn reset_corpus(root: &Path, target: &str) -> Result<PathBuf, String> {
    let dir = root.join(target);
    if dir.exists() {
        fs::remove_dir_all(&dir)
            .map_err(|error| format!("cannot reset {}: {error}", dir.display()))?;
    }
    fs::create_dir_all(&dir)
        .map_err(|error| format!("cannot create {}: {error}", dir.display()))?;
    Ok(dir)
}

fn write_seed(dir: &Path, content: &[u8]) -> Result<(), String> {
    if content.is_empty() {
        return Ok(());
    }
    fs::write(dir.join(hash(content)), content)
        .map_err(|error| format!("cannot write seed in {}: {error}", dir.display()))
}

fn hash(content: &[u8]) -> String {
    let digest = Sha1::digest(content);
    format!("{digest:x}").chars().take(16).collect()
}

fn extract_template_block(source: &str) -> Option<&str> {
    let lower = source.to_ascii_lowercase();
    let open_start = lower.find("<template")?;
    let after_name = open_start + "<template".len();
    let open_end = source[after_name..].find('>')? + after_name + 1;
    let close = lower[open_end..].find("</template>")? + open_end;
    Some(&source[open_end..close])
}
